[LOG4J2-1874] Add ByteBufferDestionation.writeBytes(ByteBuffer) and writeBytes(byte[], int, int) methods and call them from TextEncoderHelper whenever possible

leventov opened the pull request:
See description here: https://issues.apache.org/jira/browse/LOG4J2-1874

garydgregory commented:
Thank you for the PR.

Is there any way the "dangerous behaviour" can be demonstrated by a failing unit test?

leventov commented:
This PR is essentially a rebranded version of #70 which removes "bug" and "race condition" phrasing from the title and commit messages.

leventov commented:
@garydgregory this dangerous behaviour was not a bug, because it happened to always use byteBuffers which are created via `ByteBuffer.wrap(new byte[size])` or `ByteBuffer.allocate(size)`, but not `ByteBuffer.wrap(byteArray, off, len)`. But I think `ByteBuffer.array()` should always be paired with `byteBuffer.arrayOffset()` for sanity.

remkop commented:
@leventov Thanks for the PR! I will review but it will take me some time.

remkop commented:
@leventov Quick update: Started to look at the PR but did not get far yet. I'm taking my time because this is a fairly sensitive area. When converting the OutputStreamAppender and *FileAppenders to the ByteBufferDestination API for the garbage-free epic it took me a few tries to get it right. A complicating factor is that there may be users who subclassed any of these classes, and we don't want to break their stuff. At first glance your proposal does not affect that but I haven't spent enough time with it yet to be sure. (Did all the tests pass for you? I have some environment trouble and haven't been able to try yet.)

leventov commented:
> A complicating factor is that there may be users who subclassed any of these classes, and we don't want to break their stuff.

If users could implement `ByteBufferDestination`, there is a problem, because new methods added to this interface.

> Did all the tests pass for you?

Yes, except obviously unrelated failures with something about TCP/network

remkop commented:
I'm not so much concerned that users implemented ByteBufferDestination, but I'm more concerned that subclasses of OutputStreamManager and FileManager still work. For example, the new ByteBufferDestination.write(byte[], int, int) method needs to be renamed because it clashes with a pre-existing method (which was protected, not public). After that, could the existing write(byte[], int, int) method delegate to the new public one? Do they have the same concurrency semantics?

leventov commented:
@remkop renamed. Yes, OutputStreamManager.write(byte[], int, int) has the same semantics.

remkop commented:
Apologies for the delay. I now have some time to start looking at this again.

remkop commented:
I'm not done reviewing yet but I found two issues:

* StringBuilderEncoder used to have ThreadLocal<CharBuffer>, ThreadLocal<ByteBuffer> and ThreadLocal<CharsetEncoder> fields, now has one ThreadLocal<ThreadLocalState> field.  In the original code, the ThreadLocals only contains JDK classes, no custom (Log4j) classes. Where possible putting only JDK classes in ThreadLocals is preferable to avoid memory leaks in web containers: the Log4j classes may be loaded by a separate class loader which cannot be garbage collected if a thread pool threadlocal still has a reference to it. Can you revert this change?
* The build consistently hangs in this core test: testCircularSuppressedExceptions(ThrowableProxyTest.java:391) 

I tried this three times with the same result, but when building with master the build succeeds. I haven't investigated further yet. Full stack trace follows below (notice how it stops halfway a stack trace, as if a buffer is full...)

```
23:13:06.369 [main] ERROR org.apache.logging.log4j.core.impl.ThrowableProxyTest - Error
java.lang.Exception: null
        at org.apache.logging.log4j.core.impl.ThrowableProxyTest.testCircularSuppressedExceptions(ThrowableProxyTest.java:391) [test-classes/:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_131]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_131]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_131]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_131]
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50) [junit-4.12.jar:4.12]
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) [junit-4.12.jar:4.12]
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47) [junit-4.12.jar:4.12]
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325) [junit-4.12.jar:4.12]
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78) [junit-4.12.jar:4.12]
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363) [junit-4.12.jar:4.12]
        at org.junit.runners.Suite.runChild(Suite.java:128) [junit-4.12.jar:4.12]
        at org.junit.runners.Suite.runChild(Suite.java:27) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) [junit-4.12.jar:4.12]
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363) [junit-4.12.jar:4.12]
        at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55) [surefire-junit47-2.19.1.jar:2.19.1]
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137) [surefire-junit47-2.19.1.jar:2.19.1]
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107) [surefire-junit47-2.19.1.jar:2.19.1]
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83) [surefire-junit47-2.19.1.jar:2.19.1]
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75) [surefire-junit47-2.19.1.jar:2.19.1]
        at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:161) [surefire-junit47-2.19.1.jar:2.19.1]
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290) [surefire-booter-2.19.1.jar:2.19.1]
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242) [surefire-booter-2.19.1.jar:2.19.1]
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121) [surefire-booter-2.19.1.jar:2.19.1]
        Suppressed: java.lang.Exception
                at org.apache.logging.log4j.core.impl.ThrowableProxyTest.testCircularSuppressedExceptions(ThrowableProxyTest.java:392) [test-classes/:?]
                at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_131]
                at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_131]
                at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_131]
                at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_131]
                at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50) [junit-4.12.jar:4.12]
                at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) [junit-4.12.jar:4.12]
                at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47) [junit-4.12.jar:4.12]
                at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325) [junit-4.12.jar:4.12]
                at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78) [junit-4.12.jar:4.12]
                at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner.run(ParentRunner.java:363) [junit-4.12.jar:4.12]
                at org.junit.runners.Suite.runChild(Suite.java:128) [junit-4.12.jar:4.12]
                at org.junit.runners.Suite.runChild(Suite.java:27) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) [junit-4.12.jar:4.12]
                at org.junit.runners.ParentRunner.run(ParentRunner.java:363) [junit-4.12.jar:4.12]
                at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55) [surefire-junit47-2.19.1.jar:2.19.1]
                at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137) [surefire-junit47-2.19.1.jar:2.19.1]
                at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107) [surefire-junit47-2.19.1.jar:2.19.1]
                at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83) [surefire-junit47-2.19.1.jar:2.19.1]
                at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75) [surefire-junit47-2.19.1.jar:2.19.1]
                at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:161) [surefire-junit47-2.19.1.jar:2.19.1]
                at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290) [surefire-booter-2.19.1.jar:2.19.1]
                at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242) [surefire-booter-2.19.1.jar:2.19.1]
                at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121) [surefire-booter-2.19.1.jar:2.19.1]
                Suppressed: java.lang.Exception
                        at org.apache.logging.log4j.core.impl.ThrowableProxyTest.testCircularSuppressedExceptions(ThrowableProxyTest.java:391) [test-classes/:?]
                        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_131]
                        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_131]
                        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_131]
                        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_131]
                        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50) [junit-4.12.jar:4.12]
                        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) [junit-4.12.jar:4.12]
                        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47) [junit-4.12.jar:4.12]
                        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325) [junit-4.12.jar:4.12]
                        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78) [junit-4.12.jar:4.12]
                        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner.run(ParentRunner.java:363) [junit-4.12.jar:4.12]
                        at org.junit.runners.Suite.runChild(Suite.java:128) [junit-4.12.jar:4.12]
                        at org.junit.runners.Suite.runChild(Suite.java:27) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) [junit-4.12.jar:4.12]
                        at org.junit.runners.ParentRunner.run(ParentRunner.java:363) [junit-4.12.jar:4.12]
                        at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55) [surefire-junit47-2.19.1.jar:2.19.1]
                        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137) [surefire-junit47-2.19.1.jar:2.19.1]
                        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUni
```

leventov commented:
@remkop thanks for review. Replaced custom class in ThreadLocal with `Object[]`. Fixed a bug in `TextEncoderHelper`, `ThrowableProxyTest.testCircularSuppressedExceptions()` passes now.

jvz commented on log4j-core/src/main/java/org/apache/logging/log4j/core/layout/StringBuilderEncoder.java:
```diff
@@ -32,9 +32,14 @@
-    private final ThreadLocal<ByteBuffer> byteBufferThreadLocal = new ThreadLocal<>();
-    private final ThreadLocal<CharsetEncoder> charsetEncoderThreadLocal = new ThreadLocal<>();
+    /**
+     * This ThreadLocal uses raw and inconvenient Object[] to store three heterogeneous objects (CharEncoder, CharBuffer
```
So why can't you use three separate thread locals?

leventov replied:
@jvz I can. Using just one is an optimization - ThreadLocalMap is polluted less, ThreadLocalMap.get() is called only once instead of three times.

leventov replied:
Updated a comment in code

remkop commented:
@leventov Thanks for the quick turnaround. Today I only had time to check that the tests pass, but I ran `mvn clean install` twice and both times it failed here:

```
[ERROR] Tests run: 23, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.075 s <<< FAILURE! - in org.apache.logging.log4j.core.layout.PatternLayoutTest
[ERROR] testPatternSelector(org.apache.logging.log4j.core.layout.PatternLayoutTest)  Time elapsed: 0.056 s  <<< FAILURE!
java.lang.AssertionError:
Unexpected result: 2017-05-10 00:17:21,574 TRACE [main]: ====== PatternLayoutTest.testPatternSelector:259 entry ======

        at org.apache.logging.log4j.core.layout.PatternLayoutTest.testPatternSelector(PatternLayoutTest.java:261)
```

leventov commented:
@remkop this test used to depend on specific line numbers in this test. Made it more generic.

remkop commented:
Do all tests pass when you run `mvn clean install` on your machine? (may take ~20 minutes) 

    On Wednesday, May 10, 2017 12:34 AM, Roman Leventov <notifications@github.com> wrote:
 

 @remkop this test used to depend on specific line numbers in this test. Made it more generic.—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

leventov commented:
@remkop I did that for log4j-core only. Yes, it passes.

leventov commented:
@remkop any extra comments on this?

remkop commented:
I got distracted by other things but still reviewing.

leventov commented:
@remkop thanks for commitment
